src/vdaf/hits.rs

```diff
@@ -0,0 +1,741 @@
+pub trait Idpf<const KEY_LEN: usize, const OUT_LEN: usize>: Sized {
+    /// The finite field over which the IDPF is defined.
+    //
+    // NOTE(cjpatton) The IDPF of [BBCG+21] might use different fields for different levels of the
+    // prefix tree.
+    type Field: FieldElement;
+
+    /// Generate a new sequence of IDPF shares for `input`. Parameter `output` is an iterator that
+    /// is invoked to get the output value for each successive level of the prefix tree.
+    fn gen<O: IntoIterator<Item = Self::Field>>(
+        input: &IdpfInput,
+        output: O,
+    ) -> Result<[Self; KEY_LEN], VdafError>;
+
+    /// Evaluate an IDPF share on `prefix`.
+    fn eval(&self, prefix: &IdpfInput) -> Result<[Self::Field; OUT_LEN], VdafError>;
+}
+
+/// A "toy" IPDF used for demonstration purposes. The space consumed by each share is `O(2^n)`,
+/// where `n` is the length of the input. The size of each share is restricted to 1MB, so this IDPF
+/// is only suitable for very short inputs.
+//
+// NOTE(cjpatton) It would be straight-forward to generalize this construction to any `KEY_LEN` and
+// `OUT_LEN`.
+pub struct ToyIdpf<F: FieldElement> {
+    data: Vec<F>,
+    len: usize,
+}
+
+impl<F: FieldElement> Idpf<2, 1> for ToyIdpf<F> {
+    type Field = F;
+
+    fn gen<M: IntoIterator<Item = Self::Field>>(
+        input: &IdpfInput,
+        output: M,
+    ) -> Result<[Self; 2], VdafError> {
+        const MAX_DATA_BYTES: usize = 1024 * 1024; // 1MB
+
+        let max_input_len =
+            usize::try_from(log2((MAX_DATA_BYTES / F::ENCODED_SIZE) as u128)).unwrap();
+        if input.len > max_input_len {
+            return Err(VdafError::Uncategorized(format!(
+                "input length ({}) exceeds maximum of ({})",
+                input.len, max_input_len
+            )));
+        }
+
+        let data_len = 1 << (input.len + 1);
+        let mut data = vec![F::zero(); data_len];
+        let mut output = output.into_iter();
+        for len in 0..input.len + 1 {
+            data[input.prefix(len).index()] = output.next().unwrap();
+        }
+
+        // NOTE(cjpatton) We could save some space by representing one of the key shares as a PRNG
+        // seed. Of course, there's not much point, since this is a toy.
+        let mut data = split_vector(&data, 2)?.into_iter();
+        Ok([
+            ToyIdpf {
+                data: data.next().unwrap(),
+                len: input.len,
+            },
+            ToyIdpf {
+                data: data.next().unwrap(),
+                len: input.len,
+            },
+        ])
+    }
+
+    fn eval(&self, prefix: &IdpfInput) -> Result<[F; 1], VdafError> {
+        if prefix.len > self.len {
+            return Err(VdafError::Uncategorized(format!(
+                "prefix length ({}) exceeds input length ({})",
+                prefix.len, self.len
+            )));
+        }
+
+        Ok([self.data[prefix.index()]])
+    }
+}
+
+/// An input share for the heavy hitters VDAF.
```

Comment:
```suggestion
/// An input share for the heavy hitters VDAF.
#[derive(Clone, Debug, Serialize, Deserialize)]
```

Reply:
Let's hold off -- at some point, I will try to integrate this VDAF into ppm-prototype, and will probably have to make a few changes to make it all fit together. Tacking on these traits can happen then.